### ISSUE TITLE
Network::FilterChain: derived interfaces for upstream and downstream

### ISSUE THREAD

**Issue by lambdai** [envoyproxy/envoy#7560]
@kyessenov  is adding filter chain to upstream in #7503 
I am adding the functionality to drain certain downstream filter chain as improvement to LDS (also required by future FCDS). 

I prefer to add a tag to downstream filter chain as [is in listener ](https://github.com/envoyproxy/envoy/blob/37c5ce2a4df171f49148809b18afaa4adb813676/include/envoy/network/listener.h#L80) I don't see this is required by upstream filter chain.

@kyessenov  should we split UpstreamFilterChain and DownstreamFilterChain since now?

@mattklein123 WDYT?

**Comment by mattklein123** [comment]
@silentdai sorry I don't think I grok what you are asking. Can you rephrase?

**Comment by silentdai** [comment]
Sorry for the unclear and the late reply.
The background is that I am adding the functionality to drain connections owned by a subset of filter chains, instead drain all the filter chains of a listener. This functionality is the fundamental of FCDS. It can be utilized by LDS as well.

I will use LDS as example. The flow is
- A new listener config is published.
- On master thread, ListenerManager determines the filter chains need to be drained. On master thread, the filterchain is identified by FilterChain proto object
- The worker threads execute the drain. The worker has only `Envoy::Network::FilterChain` which wraps the context and now the worker cannot find which `Envoy::Network::FilterChain` need to be drained.

I am proposing add a tag() to `Envoy::Network::FilterChain`. The tag should return a int created by listener. It could be a incremental id or a hash on Filterchain proto.

I doubt if upstream FilterChain need it so I am proposing 
- `Network::FilterChain` as now
- `Network::DownstreamFilterChain` : `Network::FilterChain` { int64_t tag(); }
- `Network::upstreamFilterChain` : `Network::FilterChain` { /*nothing to add at this moment unless Kuat has plan*/ }

**Comment by silentdai** [comment]
https://github.com/silentdai/envoy/pull/4 is the example that how I would group connections by `FilterChains`

As the tag() is missing `findGroupByFilterChain` is using the address of Network::FilterChain. This highly restricted how Listener/FCDS updates the FilterChain set.

**Comment by mattklein123** [comment]
I think we can cover this in code review and can move forward in parallel with what @kyessenov is doing if that lands first.
